### PR TITLE
Dịch "rnn-scratch_vn" - Phần 6

### LINKED ISSUE #1548: Dịch "rnn-scratch_vn" - Phần 6
Phần cần dịch: https://github.com/aivivn/d2l-vn/blame/master/chapter_recurrent-neural-networks/rnn-scratch_vn.md#L444-L516

### Hướng dẫn
- [Dịch trực tiếp trên website](https://github.com/aivivn/d2l-vn/blob/master/CONTRIBUTING.md).
- Sau khi nhấn vào phần cần dịch ở trên, hãy chọn **Normal view** và làm theo hướng dẫn ở trên để có thể bắt đầu dịch.

### PR DESCRIPTION
**Pull request by nguyenduydu**
## Pull Request Checklist

**Đánh dấu tick (sau khi "Create Pull Request") vào các mục dưới đây:**

* [x] Pull Request này tương ứng với issue nào? Trả lời số issue sau dấu `#` (không có dấu cách): Close #1548

* [x] Bản dịch này có bám sát **[Bảng thuật ngữ](https://github.com/aivivn/d2l-vn/blob/master/glossary.md)** không? Nếu một từ/cụm chưa có trong Bảng thuật ngữ, hãy tạo một Pull Request riêng đề xuất cách dịch từ/cụm đó.

* [x] Format của bản gốc có được giữ nguyên không?
    * Các phần in nghiêng, in đậm.
    * Tiêu đề (số lượng dấu `#` đầu dòng).
    * Bảng biểu, chú thích cho bảng (dòng phía trên mỗi bảng bắt đầu bằng dấu `:`).
    * Dịch các từ trong hình vẽ nếu cần. Các dòng có hình có dạng: `caption`.
    * Dịch các chú thích hình vẽ (thay các cụm `*dịch chú thích ảnh phía trên*` bằng bản dịch tiếng Việt).
    * Không dịch các phần code (nằm giữa hai dấu `````).
    * Copy các công thức toán từ bản gốc (các đoạn có `$`).
    * Giữ các dòng gán nhãn (bắt đầu với `:label:`).
    * Không tự thêm bớt các dòng trắng.
     
* [x] Trong một đoạn văn, mỗi câu văn đã được viết trong một dòng, giữa các dòng không có dòng trắng. Markdown sẽ coi những dòng liền nhau không có dòng trắng là một đoạn văn. Việc này giúp công đoạn review được thuận tiện hơn.

* [x] Tất cả chữ trong các hình vẽ đã được dịch. Nếu chưa, vui lòng xem thêm **[hướng dẫn hoặc yêu cầu dịch hình](https://github.com/aivivn/d2l-vn/blob/master/img/README.md)**.

* [x] Điền tên của bạn vào mục "Những người thực hiện" ở cuối file.

* [x] **Sau khi được approve**, điền tên những reviewers mà bạn thấy có nhiều đóng góp cho Pull Request này. Xem danh sách Contributors **[tại đây](https://github.com/aivivn/d2l-vn/blob/master/docs/contributors_info.md)**.

## Với Reviewer

1. Chỉ review nếu Pull Request này đã sẵn sàng cho việc review.

2. Đã đọc kỹ hướng dẫn đóng góp và xem qua **[Bảng thuật ngữ](https://github.com/aivivn/d2l-vn/blob/master/glossary.md)**.

3. **Bạn có thể đề xuất thay đổi cách dịch mỗi dòng trực tiếp như trong hình dưới đây:**
img

### CODE REVIEW

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 1]
```diff
@@ -471,7 +472,7 @@ train_ch8(model, train_iter, vocab, lr, num_epochs, ctx, use_random_iter=True)
 While implementing the above RNN model from scratch is instructive, it is not convenient. In the next section we will see how to improve significantly on the current model and how to make it faster and easier to implement.
 -->
 
-*dịch đoạn phía trên*
+Mặc dù việc lập trình mô hình RNN từ đầu có tính hướng dẫn, nhưng nó không thực sự tiện lợi. Trong phần tiếp theo, ta sẽ xem cách cải thiện mô hình hiện tại một cách đáng kể để làm cho nó nhanh hơn và dễ triển khai hơn.
```
Comment (lkhphuc) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:475]:
```suggestion
Mặc dù ta đã học được nhiều từ việc lập trình mô hình RNN từ đầu nhưng cách làm này không thực sự tiện lợi lắm. 
Trong phần tiếp theo, ta sẽ xem cách cải thiện mô hình hiện tại một cách đáng kể và cách làm nó nhanh và dễ lập trình hơn.
```

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 2]
```diff
@@ -489,7 +490,12 @@ While implementing the above RNN model from scratch is instructive, it is not co
 * Sequential partitioning typically leads to better models.
 -->
 
-*dịch đoạn phía trên*
+* Mô hình chuỗi cần khởi tạo trạng thái cho quá trình huấn luyện.
+* Giữa các mô hình chuỗi, ta cần đảm bảo tách các gradient, để chắc chắn rằng phép vi phân tự động không lan truyền các hiệu ứng ra ngoài mẫu hiện tại.
```
Comment (lkhphuc) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:494]:
```suggestion
* Giữa các mô hình chuỗi, ta cần đảm bảo tách các gradient, để chắc chắn rằng phép vi phân tự động không lan truyền các ảnh hưởng ra ngoài mẫu hiện tại.
```

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 3]
```diff
@@ -511,7 +517,17 @@ While implementing the above RNN model from scratch is instructive, it is not co
...
-*dịch đoạn phía trên*
+1. Chứng minh rằng biễu diễn one-hot tương đương với việc chọn một embedding khác nhau cho từng đối tượng.
+2. Điều chỉnh các siêu tham số để cải thiện độ rối rắm.
+    * Bạn có thể giảm nó xuống bao nhiêu? Thay đổi embedding, số nút ẩn, tốc độ học, vv
+    * Mô hình này sẽ hoạt động như thế nào trên các cuốn sách khác của H. G. Wells, ví dụ như [The War of the Worlds] (http://www.gutenberg.org/ebooks/36).
+3. Thay đổi hàm dự đoán sử dụng phép lấy mẫu thay vì chọn ký tự tiếp theo có khả năng cao nhất.
+    * Điều gì sẽ xảy ra?
+    * Điều chỉnh mô hình để ưu tiên các đầu ra có khả năng cao hơn, ví dụ: bằng cách lấy mẫu từ $q(w_t \mid w_{t-1}, \ldots, w_1) \propto p^\alpha(w_t \mid w_{t-1}, \ldots, w_1)$ for $\alpha > 1$.
+4. Điều gì sẽ xảy ra nếu ta chạy mã nguồn trong phần này mà không xén gradient?
+5. Thay đổi phép lấy mẫu liền kề để các trạng thái ẩn không bị tách khỏi biểu đồ tính toán. Thời gian chạy và độ chính xác có thay đổi không?
```
Comment (minhduc0711) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:528]:
```suggestion
5. Thay đổi phép lấy mẫu liền kề để các trạng thái ẩn không bị tách khỏi đồ thị tính toán. Thời gian chạy và độ chính xác có thay đổi không?
```

**Comment by nguyenduydu** [comment]
Cám ơn @lkhphuc  và @minhduc0711  đã góp ý nhé!

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 4]
```diff
@@ -460,7 +461,7 @@ train_ch8(model, train_iter, vocab, lr, num_epochs, ctx)
 Finally let us check the results to use a random sampling iterator.
 -->
 
-*dịch đoạn phía trên*
+Cuối cùng, ta sẽ kiểm tra kết quả khi sử dụng một iterator để lấy mẫu ngẫu nhiên.
```
Comment (lkhphuc) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:464]:
```suggestion
Cuối cùng, ta sẽ kiểm tra kết quả khi sử dụng một bộ lặp để lấy mẫu ngẫu nhiên.
```
uả sao chỗ này thấy @minhduc0711 sửa lại iterator nhỉ?

Comment (rootonchair) [reply chapter_recurrent-neural-networks/rnn-scratch_vn.md:464]: @lkhphuc tại vì mình đã thống nhất giữ nguyên á anh
https://github.com/aivivn/d2l-vn/blob/master/glossary.md#i

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 5]
```diff
@@ -489,7 +491,12 @@ While implementing the above RNN model from scratch is instructive, it is not co
 * Sequential partitioning typically leads to better models.
 -->
 
-*dịch đoạn phía trên*
+* Mô hình chuỗi cần khởi tạo trạng thái cho quá trình huấn luyện.
+* Giữa các mô hình chuỗi, ta cần đảm bảo tách các gradient, để chắc chắn rằng phép vi phân tự động không lan truyền các ảnh hưởng ra ngoài mẫu hiện tại.
```
Comment (rootonchair) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:495]:
```suggestion
* Giữa các mô hình chuỗi, ta cần đảm bảo tách các gradient, để chắc chắn rằng phép vi phân tự động không lan truyền các ảnh hưởng ra ngoài phạm vi mẫu hiện tại.
```

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 6]
```diff
@@ -489,7 +491,12 @@ While implementing the above RNN model from scratch is instructive, it is not co
...
+* Mô hình ngôn ngữ RNN đơn giản bao gồm một bộ mã hóa, một mô hình RNN và một bộ giải mã.
+* Gọt gradient có thể ngăn sự bùng nổ gradient nhưng không thể khắc phục được vấn đề tiêu biến gradient.
+* Độ rối rắm đánh giá hiệu năng của mô hình trên các chuỗi có độ dài khác nhau. Nó là trung bình lũy thừa của mất mát entropy chéo.
```
Comment (rootonchair) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:498]:
```suggestion
* Độ rối rắm đánh giá chất lượng mô hình trên các chuỗi có độ dài khác nhau. Nó là trung bình lũy thừa của mất mát entropy chéo.
```
https://github.com/aivivn/d2l-vn/blob/master/glossary.md#p

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 7]
```diff
@@ -511,7 +518,17 @@ While implementing the above RNN model from scratch is instructive, it is not co
 7. Prove that the perplexity is the inverse of the harmonic mean of the conditional word probabilities.
 -->
 
-*dịch đoạn phía trên*
+1. Chứng minh rằng biễu diễn one-hot tương đương với việc chọn một embedding khác nhau cho từng đối tượng.
+2. Điều chỉnh các siêu tham số để cải thiện độ rối rắm.
+    * Bạn có thể giảm nó xuống bao nhiêu? Hãy thay đổi embedding, số nút ẩn, tốc độ học, vv
+    * Mô hình này sẽ hoạt động tốt đến đâu trên các cuốn sách khác của H. G. Wells, ví dụ như [The War of the Worlds] (http://www.gutenberg.org/ebooks/36).
+3. Thay đổi hàm dự đoán sử dụng phép lấy mẫu thay vì chọn ký tự tiếp theo có khả năng cao nhất.
```
Comment (rootonchair) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:525]:
```suggestion
3. Thay đổi hàm dự đoán bằng việc lấy mẫu thay vì chọn ký tự tiếp theo là ký tự có khả năng cao nhất.
```

**Comment by nguyenduydu** [comment]
Cám ơn @rootonchair đã góp ý nhé.

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 8]
```diff
@@ -448,7 +448,8 @@ Now we can train a model.
 Since we only use $10,000$ tokens in the dataset, the model needs more epochs to converge.
 -->
 
-*dịch đoạn phía trên*
+Bây giờ thì ta có thể huấn luyện mô hình.
+Do chỉ sử dụng $10,000$ token trong tập dữ liệu, mô hình này sẽ cần nhiều epoch hơn để hội tụ.
```
Comment (duythanhvn) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:452]:
```suggestion
Do chỉ sử dụng $10.000$ token trong tập dữ liệu, mô hình này sẽ cần nhiều epoch hơn để hội tụ.
```
Từ bây giờ mình dùng dấu `.` thay vì dấu `,` trong các trường hợp số học này ở bản dịch tiếng Việt của mình nghen.

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 9]
```diff
@@ -511,7 +518,17 @@ While implementing the above RNN model from scratch is instructive, it is not co
 7. Prove that the perplexity is the inverse of the harmonic mean of the conditional word probabilities.
 -->
 
-*dịch đoạn phía trên*
+1. Chứng minh rằng biễu diễn one-hot tương đương với việc chọn một embedding khác nhau cho từng đối tượng.
+2. Điều chỉnh các siêu tham số để cải thiện độ rối rắm.
+    * Bạn có thể giảm nó xuống bao nhiêu? Hãy thay đổi embedding, số nút ẩn, tốc độ học, vv
```
Comment (duythanhvn) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:523]:
```suggestion
    * Bạn có thể giảm nó xuống bao nhiêu? Hãy thay đổi embedding, số nút ẩn, tốc độ học, vv
```

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 10]
```diff
@@ -511,7 +518,17 @@ While implementing the above RNN model from scratch is instructive, it is not co
...
+    * Mô hình này sẽ hoạt động tốt đến đâu trên các cuốn sách khác của H. G. Wells, ví dụ như [The War of the Worlds] (http://www.gutenberg.org/ebooks/36).
```
Comment (duythanhvn) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:524]:
```suggestion
    * Mô hình này sẽ hoạt động tốt đến đâu trên các cuốn sách khác của H. G. Wells, ví dụ như [The War of the Worlds] (http://www.gutenberg.org/ebooks/36).
```

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 11]
```diff
@@ -511,7 +518,17 @@ While implementing the above RNN model from scratch is instructive, it is not co
...
+3. Thay đổi hàm dự đoán bằng việc lấy mẫu thay vì chọn ký tự tiếp theo là ký tự có khả năng cao nhất.
+    * Điều gì sẽ xảy ra?
```
Comment (duythanhvn) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:526]:
```suggestion
    * Điều gì sẽ xảy ra?
```

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 12]
```diff
@@ -511,7 +518,17 @@ While implementing the above RNN model from scratch is instructive, it is not co
...
+    * Điều chỉnh mô hình để ưu tiên các đầu ra có khả năng cao hơn, ví dụ: bằng cách lấy mẫu từ $q(w_t \mid w_{t-1}, \ldots, w_1) \propto p^\alpha(w_t \mid w_{t-1}, \ldots, w_1)$ với $\alpha > 1$.
```
Comment (duythanhvn) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:527]:
```suggestion
    * Điều chỉnh mô hình để ưu tiên các đầu ra có khả năng cao hơn, ví dụ: bằng cách lấy mẫu từ $q(w_t \mid w_{t-1}, \ldots, w_1) \propto p^\alpha(w_t \mid w_{t-1}, \ldots, w_1)$ với $\alpha > 1$.
```
